docs/specs/sections/implementation.tex

```diff
@@ -0,0 +1,89 @@
+
+\subsection{Participants} 
+Protocol implementation involves realization of the protocol building blocks as well as providing means of data communication between them. Building blocks are placed at the following locations, which correspond to protocol participants:
+
+\begin{itemize}%[label=$\bullet$]
+	\item User software.
+	\item License Provider software.
+	\item Service Provider software.
+	\item Citadel contract.
+\end{itemize}
+
+\begin{flushleft}
+Data communication between protocol participants is realized in the following communication modes:
+\end{flushleft}
+
+\begin{itemize}%[label=$\bullet$]
+	\item Transactions which change the contract state.
+	\item Queries which do not change the contract state.
+	\item Storing data directly in blockchain by issuing transactions.
+	\item Retrieving data from blockchain by scanning transactions.
+	\item Delegating ZK proof calculation.
+	\item Off-chain calls.
+\end{itemize}
+
+\begin{flushleft}
+The following diagram illustrates an interaction between protocol participants and indicates the communication means used.
+\end{flushleft}
+
+\begin{figure}[h]
+	\centering
+		\includegraphics[width=390pt,draft=false]{images/implementation.png}
+	\caption{Interaction between protocol participants}
+	\label{fig:implementation}
+\end{figure}
+
+\begin{flushleft}
+On the diagram, we can see various communication modes being used. Initially, user submits recipient-less transaction which contains a request as payload into the blockchain. Subsequently, License Provider can scan the blockchain for transactions containing relevant requests. License Provider can obtain requests via other routes as well, for example via http or email, passing requests on blockchain is only one of possible ways. Once License Provider gets a hold of a request, it can perform appropriate verification and issue a license. Issuing a license involves another mode of communication - a smart contract transaction call. Smart contract transaction call is also a blockchain transaction, yet to not to confuse the reader, we show it on the diagram while omitting the detail of blockchain  involvement. User obtains licenses using a mode of communication based on a contract query. For privacy reasons, user obtains a bulk of information not pertaining exclusively to her, and she filters it out by herself. To economize the volume of data transfers, block-height range is also passed to allow the transfer to include only a subset of available records. All modes of communication used so far were on-chain. Communication between user and ServiceProvider, on the other hand, is off-chain. When Service Provider wants to establish a session, it calls a contract to obtain session for a given session id.
```

Comment:
All agreed except for the last one - it must be `a session` because it is not a concrete session, but rather a new, not yet known and not yet introduced session. If it was a session id, I'd say `the session id` but for session, I'd say `a session`. Internet seems to confirm that.